### ISSUE TITLE
RFC: Managing dependencies across packages

### ISSUE THREAD

**Issue by Borales** [deity-io/falcon#654]
# Background

We've been having this problem for a long time now. Whenever a new package version (from the dependencies of our Falcon packages) comes up - end developers would have to wait until we make an upgrade and release a newer version as well (and this may take some time). 

Sometimes such releases may contain important performance/security fixes, which have to be applied immediately. Our goal is to make Falcon implementation more easy and smooth.

We also need to make sure we minimize the possibility of having the same package installed with multiple versions, which may lead to unexpected behavior.

# Proposed solution

We should be more "open" (using `^` for the dependency version) for common packages. As practice shows - we cannot make a single decision for all dependencies and follow it blindly, thus not every package can be used with `^` version, as it may take some work to be done on Falcon side to adopt a newer package version (like babel/webpack dependencies - our code has to know how to use those new things and not to break other stuff), so some things will remain the same for some of them.

### Existing internal convention

> `dependencies` - all pinned, except `@deity/*` packages
> `devDependencies` - all pinned, except `@deity/*` packages
> `peerDependencies` - semantic

Original source - https://github.com/deity-io/falcon/pull/627#discussion_r332925161

### Proposed convention

**General suggestion**

When using in `devDependencies` only
- we should consider using `~` (latest `patch` version available) to cover important patches for build dependencies (like babel/webpack/rollup) considering such patch version won't bring any new features or breaking changes.
- for `@types/*` - we could use `~` as well (to keep them up to date).
- the rest of packages - could be pinned

`peerDependencies` + `devDependencies`:
- Use `^` to make sure we the provided version range is being tested correctly when using on the target project

**Overall summary:**
1. When a package is required for testing purposes only - it goes to `devDependencies`.
2. When a package is required on the target application and a developer's choice is also **required** - it goes to `devDependencies` + `peerDependencies` (with the same version range, totally controlled by a developer).
3. When a package is required on the target application and a developer's choice is **optional** - it goes to `dependencies` with `^` (may be controlled by a developer, but by default - it's controlled by us).
4. When a package is required on the target application and a developer's choice is **not allowed** - it goes to `dependencies` with a pinned version (controlled only by us).

---

**Specific examples**

FalconServer:
- `@deity/falcon-server` package could be considered as the _main application_ since FS-based project does not contain that much code by default, considering APIs and Extensions will be delivered via project dependencies, so main FS dependencies can remain in `dependencies` section, all FS-based app needs is a single `@deity/falcon-server` dependency and `index.js` entry file importing this package to start.
- All shared packages (like `graphql`, `graphql-tools`, `apollo-*`) should be delivered by `@deity/falcon-server` (`dependencies`) with `^` symbol, **providing a way of pinning a specific version in the FS-based app dependencies and leaving this possibility to a developer**
- FS-related packages (like `falcon-server-env`) can count on having these dependencies installed within the project (from FS), so they should have them in their `devDependencies` + `peerDependencies`.

FalconClient:
- In case with FC - we know that some packages are required to be defined in the project's dependencies in order to build it (like `react` packages). These could be defined in `peerDependencies` section. 
- The usage of `^` remains the same as described above.
- As a precaution for some cases - we could consider specifying the upper supported version.
- For instance, in our current example app, `client/package.json` has `"apollo-client": "2.6.4"` dependency, as well as FC has the same one. When FC is gonna update the version, the example app would have to use the same version (in order to keep a single instance of `apollo-client` installed in the project), so a developer **would have to** check `apollo-client` version in our package and bump the version manually each time. Having `"apollo-client": "^2.6.4"` in FC - **we leave this choice to a developer** (which could also set `^` and stop worrying about manual updates and let NPM control it), this way we reduce a chance of having installed the same package with multiple versions.

---

# Definition of done

- [ ] Agree on a certain strategy of using/managing dependencies across Falcon packages and Falcon-based apps 
- [ ] Create a PR to https://github.com/deity-io/falcon-docs to write down the approved strategy
- [ ] Make sure packages (and example app) have `package.json` files are aligned with the agreed strategy (fix `dependencies`/`devDependencies`/`peerDependencies` entries)

cc @deity-io/core-developers-js

---

_(this placeholder will be replaced when some decisions are made)_

**Comment by nimobeeren** [comment]
I agree on making important build dependencies at least `~`. I probably don't understand the reasons, but I would even go further and say all external dependencies should be at least `~`, maybe even `^`. As I understand, minor versions should not include breaking changes. However, it is important that in the case that one of the dependencies *accidentally* breaks something, Falcon as a whole does not break. So as a compromise I would say `~` for all external packages is safe enough.

I've looked at [Next.js package.json](https://github.com/zeit/next.js/blob/master/packages/next/package.json) as an example, and they do used pinned versions for all dependencies.

This is a tough issue and requires more effort to get the details right.
